### ISSUE TITLE
create test data file

### ISSUE THREAD

**Issue by razy** [deepanwayx/Sentence-Classification-With-Convolutional-Neural-Networks#1]
Hi Deepanwayx, thanks for posting this awesome work. I have learned a lot. how to modify the source code to test the data file if I want to make five classes as positive, very positive, neutral, negative and very negative. I want took 100 examples from each of them to create the test.data file. The first of test.data whether the data should be negative sentiment followed the positive sentiment or can randomly ??

how modification of this section for five classes:
y_test = np.array (list (np.zeros (50)) + list (np.ones (50)))
y_prediction = np.zeros ((10, 100))

regards.

**Comment by deepanwayx** [comment]
There were a few typos, so I have updated the repository. Consider running the new files if you want to replicate results.

In five class classification problem, the files you need to have are 'rt-polarity.veryneg', ''rt-polarity.neg', 'rt-polarity.neutral', 'rt-polarity.pos' and 'rt-polarity.verypos'. And as you said 'test.data' containing 100 examples from each of them in the order: very_neg, ....... , very_pos. The order is not really important as long as if you have a csv file or dataframe containing the sentiments you can read from to create y_test. . Here for the sake of demonstration we have put the sentences in those order, so that we can easily create y_test.  

Assuming very_neg to be 0 and very_pos to be 4 do the following changes :
In 'conv_net_sentence.ipynb' 
from

`y_test = np.array(list(np.zeros(50)) + list(np.ones(50)))   # the actual test classes`

to

`y_test = np.array(list(np.zeros(100)) + list(np.ones(100)) + list(2*np.ones(100)) + list(3*np.ones(100)) + list(4*np.ones(100)))  # the actual test classes`

and
from

`y_prediction = train_conv_net(..... , hidden_units=[16,2], ......)`

to 

`y_prediction = train_conv_net(..... , hidden_units=[16,5], ......)`

A few modifications will also have to be made in 'process_data.ipynb' which I have done for you in 'process_data_5_classes.ipynb'. See this [https://github.com/deepanwayx/Sentence-Classification-With-Convolutional-Neural-Networks/blob/master/process_data_5_classes.ipynb](url)

This should work for you. Let me know if there is any problem.

As a sidenote I also suggest you to check this out because in many cases you will have much more than five classes and it is almost near impossible to list all the files by hand and read them, so you will need a dataframe containing the features and labels. [https://github.com/deepanwayx/Ensemble-Word-Embedding-Stanford-Sentiment-Treebank-1/blob/master/Autoencoder-Ensemble/process_data%20I.ipynb](url)

**Comment by razy** [comment]
Thank you for your help...
I've tried it and it works ...
and the result is : 

`[('image shape', 313, 300), ('filter shape', [(16, 1, 3, 300), (16, 1, 4, 300), (16, 1, 5, 300)]), ('hidden_units', [16, 5]), ('dropout', [0.5]), ('batch_size', 50), ('non_static', True), ('learn_decay', 0.95), ('conv_non_linear', 'relu'), ('non_static', True), ('sqr_norm_lim', 9), ('shuffle_batch', True)]

... training
epoch: 1, training time: 478.60 secs, train perf: 30.68 %, val perf: 27.74 %
epoch: 2, training time: 472.46 secs, train perf: 39.87 %, val perf: 32.97 %
Test Accuracy  : 27.8 %`

maybe I should do the epoch more than 2 times to improve accuracy.
there are some things I want to ask.

1. `[( 'Image shape', 64, 300)]` shape image by default is 64, in my case automatically change to 313, does this matter? what does it mean figure of 300 after the number 64?

2. then the `'filter shape', [(16, 1, 3, 300), (16, 1, 4, 300), (16, 1, 5, 300)])` what does it mean the numbers 16 and 1?

3. The original author did not used test.data, so for testing the data used data Where?

4. Where the data are used to train perf? whether taken from the training data randomly?

I tried to open the link you gave [https://github.com/deepanwayx/Ensemble-Word-Embedding-Stanford-Sentiment-Treebank-1/blob/master/Autoencoder-Ensemble/process_data%20I.ipynb](url) but the page can not be found

sorry for asking too much. Thank you so much for your help.
regards.

**Comment by deepanwayx** [comment]
@razy 

1 . To represent all the sentences in train, test data in vectors of same length and to make the convolution filters work appropriately we need to pad the sentences at the beginning and at the end. Maximum length of sentence in my case was 56. As maximum length of filter is 5. So each sentence is
represented by vectors of length 56 + 2*(5-1) = 64. Your sentences must be longer, so you are getting 313. 
Pretrained Word2Vec and GloVe represents each word by a vector of 300 dimension which is printed in 'Image Shape'.

2 . After all the convolution layers you have a fully connected layer, in other words a simple neural network. The number of units in the hidden layer is 16. You are free to change this here, but additional units come with additional training time.

    `y_prediction = train_conv_net(..... , hidden_units=[16,2], ......)`

If X is a matrix of r rows and c columns you can always make the 2D Matrix a 3D tensor with r rows,         unit depth and c columns. The 1 after the 16 is actually representing the unit depth. Much like 2D images represented in 3D tensors as RGB channels. You can refer to theano documentation for more                             information about tensors.

3 . The original author did not used test data, instead he used a 10 fold cross validation over the training data. The task he showed was subjectivity classification. But in the original paper there were some other tasks also which had explicit test dataset. I created the repo to see how CNN for sentence works for the other tasks.

4 . In 'process_data_5_classes.ipynb' there is this part of code:

`def build_data_cv(data_folder, cv=10, clean_string=True):
  ......................................
            datum  = {"y":1, 
                      "text": orig_rev,                             
                      "num_words": len(orig_rev.split()),
                      "split": np.random.randint(0,cv)}
            train_revs.append(datum)
.................................`

At this time training and validation data both are in train_revs and the function assigns a random number between 0 and 9 to each sentence. Now in 'conv_net_sentence.ipynb' when you do:

`r = range(0,10)    
for i in r:
    datasets = make_idx_train_valid_data(train_revs, word_idx_map, i, max_l = 56, k = 300, filter_h = 5)`

in first epoch valid data becomes those sentence which has that random number 0, rest become training data, in second epoch valid data become those sentences which has that random number 1, rest become training data and so on.
In my code I actually did this for time constraint
`r = range(0,5)    
for i in r:
    datasets = make_idx_train_valid_data(train_revs, word_idx_map, i, max_l = 56, k = 300, filter_h = 5)`

As np.random.randint is a pseudo random number generator so we can expect that each number between 0-9 will get assigned to almost same number of sentences.

As for the link address try copying it and pasting in another tab.

**Comment by razy** [comment]
Thank you for the explanation. I became more understanding now.
after I increased the number of epoch, its accuracy is better .
`epoch: 1, training time: 241.18 secs, train perf: 49.52 %, val perf: 45.80 %`
`epoch: 2, training time: 244.04 secs, train perf: 79.60 %, val perf: 77.63 %`
`epoch: 3, training time: 247.30 secs, train perf: 84.79 %, val perf: 81.79 %`
`epoch: 4, training time: 248.90 secs, train perf: 87.43 %, val perf: 82.81 %`
`epoch: 5, training time: 249.36 secs, train perf: 89.48 %, val perf: 84.07 %`
`epoch: 6, training time: 248.72 secs, train perf: 91.10 %, val perf: 84.52 %`
`epoch: 7, training time: 245.67 secs, train perf: 92.29 %, val perf: 84.52 %`
`Test Accuracy   : 70.8 %`
